cpp/test/sparse/neighbors/connect_components.cu

```diff
@@ -353,5 +360,612 @@ TEST_P(ConnectComponentsTestF_Int, Result)
+    thrust::scatter(raft::resource::get_thrust_policy(handle),
+                    core_dists,
+                    core_dists + m,
+                    map,
+                    tmp_core_dists.data_handle());
+    raft::copy_async(
+      core_dists, tmp_core_dists.data_handle(), m, raft::resource::get_cuda_stream(handle));
+  }
+};
+
+template <typename value_t, typename value_idx>
+struct ConnectComponentsMutualReachabilityInputs {
+  value_idx n_row;
+  value_idx n_col;
+  std::vector<value_t> data;
+  std::vector<value_t> core_dists;
+  std::vector<value_idx> colors;
+  std::vector<value_idx> expected_rows;
+  std::vector<value_idx> expected_cols;
+  std::vector<value_t> expected_vals;
+};
+
+template <typename value_idx, typename value_t>
+class ConnectComponentsEdgesTest
+  : public ::testing::TestWithParam<ConnectComponentsMutualReachabilityInputs<value_t, value_idx>> {
+ protected:
+  void basicTest()
+  {
+    raft::resources handle;
+
+    auto stream = resource::get_cuda_stream(handle);
+
+    params = ::testing::TestWithParam<
+      ConnectComponentsMutualReachabilityInputs<value_t, value_idx>>::GetParam();
+
+    raft::sparse::COO<value_t, value_idx> out_edges(resource::get_cuda_stream(handle));
+
+    rmm::device_uvector<value_t> data(params.n_row * params.n_col,
+                                      resource::get_cuda_stream(handle));
+    rmm::device_uvector<value_t> core_dists(params.n_row, resource::get_cuda_stream(handle));
+    rmm::device_uvector<value_idx> colors(params.n_row, resource::get_cuda_stream(handle));
+
+    raft::copy(data.data(), params.data.data(), data.size(), resource::get_cuda_stream(handle));
+    raft::copy(core_dists.data(),
+               params.core_dists.data(),
+               core_dists.size(),
+               resource::get_cuda_stream(handle));
+    raft::copy(
+      colors.data(), params.colors.data(), colors.size(), resource::get_cuda_stream(handle));
+
+    /**
+     * 3. connect_components to fix connectivities
+     */
+    MutualReachabilityFixConnectivitiesRedOp<value_idx, value_t> red_op(core_dists.data(),
+                                                                        params.n_row);
+
+    raft::linkage::connect_components<value_idx, value_t>(
+      handle, out_edges, data.data(), colors.data(), params.n_row, params.n_col, red_op, 13, 1);
+
+    ASSERT_TRUE(
+      devArrMatch(out_edges.rows(), params.expected_rows.data(), out_edges.nnz, Compare<int>()));
+
+    ASSERT_TRUE(
+      devArrMatch(out_edges.cols(), params.expected_cols.data(), out_edges.nnz, Compare<int>()));
+
+    ASSERT_TRUE(devArrMatch(
+      out_edges.vals(), params.expected_vals.data(), out_edges.nnz, CompareApprox<float>(1e-4)));
+  }
+
+  void SetUp() override { basicTest(); }
+
+  void TearDown() override {}
+
+ protected:
+  ConnectComponentsMutualReachabilityInputs<value_t, value_idx> params;
+};
+
+const std::vector<ConnectComponentsMutualReachabilityInputs<float, int>> mr_fix_conn_inputsf2 = {
```

Comment:
I don't think that hardcoding a test case like this is a good idea. Is there a way to generate a test case in a shorter way?

Reply:
The reason it was hardcoded is that the test was written by running `HDBSCAN` on a custom `make_blobs` dataset and printing (and then manually copy-pasting) the different arrays needed (`X`, `colors`, `core_dists`) and comparing that the edges are the same as the ones generated by fused NN.

Reply:
I do agree that we should be striving to automate the data generation and find reasonable assertions that exercise the behavioral characteristics of the algorithms where possible without resorting to brittle hardcoded assertions. One of the big challenges in this algorithm is that it builds upon so many core pieces that it, in and of itself, is pretty end-to-end and it's hard to mock up a quick "naive" version without re-writing an entire algorithm on the CPU. 

I like @ahendriksen's idea of testing some of the edge cases here, especially since these changes do affect so much of the core logic in multiple algorithms which are being heavily used at the moment. We should do our diligence and write additional assertions to make sure we're not introducing silly bugs that weren't a problem in the prior version. 

@tarang-jain, normally I would say let's worry about this later, but this being such a core primitive, I think we should try to capture at least the behaviors described by @ahendriksen above, so we can all feel confident that this is really as solid as it looks.